constraint-solver/src/solver/linearizer.rs

```diff
@@ -331,11 +131,29 @@ impl<T: RuntimeConstant + Hash, V: Clone + Eq + Ord + Hash> Linearizer<T, V> {
                 .map(|var| GroupedExpression::from_unknown_variable(var.clone()))
         }
     }
+
+    /// Returns an iterator over expressions equivalent to `expr` with the idea that
+    /// they might allow to answer a query better or worse.
+    /// It usually returns the original expression, a single variable that it was
+    /// substituted into during a previous linearization and a previously linearized version.
+    pub fn internalized_versions_of_expression(
```

Comment:
I think it's kind of complicated to explain this function and why it's needed. It's only needed for `range_constraint_for_expression`. Maybe this function could be be implemented directly (receiving an `&impl RangeConstraintProvider`)?